src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp

```diff
@@ -0,0 +1,301 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ *
+ */
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/microwave-oven-control-server/microwave-oven-control-server.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/error-mapping.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::MicrowaveOvenControl;
+using namespace chip::app::Clusters::MicrowaveOvenControl::Attributes;
+using OperationalStateEnum = OperationalState::OperationalStateEnum;
+using Status               = Protocols::InteractionModel::Status;
+
+namespace {
+
+constexpr uint32_t kMaxCookTime     = 65535;
+constexpr uint32_t kMinCookTime     = 1;
+constexpr uint8_t kDefaultCookTime  = 30;
+constexpr uint8_t kDefaultCookPower = 100;
+
+constexpr size_t kMicrowaveOvenControlInstanceTableSize =
+    EMBER_AF_MICROWAVE_OVEN_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT;
+
+static_assert(kMicrowaveOvenControlInstanceTableSize <= kEmberInvalidEndpointIndex,
+              "Microwave Oven Control Instance table size error");
+
+OperationalState::Instance * gInstanceTable[kMicrowaveOvenControlInstanceTableSize] = { nullptr };
+
+} // anonymous namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace MicrowaveOvenControl {
+
+/**
+ * Get operational state instance
+ */
+OperationalState::Instance * GetOPInstance(EndpointId aEndpoint)
+{
+    uint16_t ep = emberAfGetClusterServerEndpointIndex(aEndpoint, MicrowaveOvenControl::Id,
+                                                       EMBER_AF_MICROWAVE_OVEN_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT);
+    return (ep >= kMicrowaveOvenControlInstanceTableSize ? nullptr : gInstanceTable[ep]);
+}
+
+/**
+ * Set operational state instance
+ */
+void SetOPInstance(EndpointId aEndpoint, OperationalState::Instance * aInstance)
+{
+    uint16_t ep = emberAfGetClusterServerEndpointIndex(aEndpoint, MicrowaveOvenControl::Id,
+                                                       EMBER_AF_MICROWAVE_OVEN_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT);
+    // if endpoint is found
+    if (ep < kMicrowaveOvenControlInstanceTableSize)
+    {
+        gInstanceTable[ep] = aInstance;
+    }
+}
+
+bool CheckCookTimeIsInRange(uint32_t cookTime)
```

Comment:
Suggest `IsCookTimeInRange` and `IsPowerSettingInRange` --> read better (`if(IsCookTimeInRange(123)`).